base_layer/core/src/chain_storage/blockchain_database.rs

```diff
@@ -0,0 +1,259 @@
+// following conditions are met:
+//
+// 1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following
+// disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the
+// following disclaimer in the documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote
+// products derived from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+// INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+// USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+
+use crate::{
+    blocks::{block::Block, blockheader::BlockHeader},
+    chain_storage::{
+        error::ChainStorageError,
+        transaction::{DbKey, DbTransaction, DbValue, MetadataKey, MetadataValue},
+        ChainMetadata,
+    },
+    proof_of_work::Difficulty,
+    transaction::{TransactionKernel, TransactionOutput},
+    types::HashOutput,
+};
+use log::*;
+use std::sync::{Arc, RwLock, RwLockReadGuard};
+
+const LOG_TARGET: &str = "core::chain_storage::database";
+
+pub enum MmrTree {
+    Utxo,
+    Kernel,
+    RangeProof,
+    Header,
+}
+
+pub trait BlockchainBackend: Send + Sync {
+    fn write(&self, tx: DbTransaction) -> Result<(), ChainStorageError>;
+    fn get(&self, key: &DbKey) -> Result<Option<DbValue>, ChainStorageError>;
+    fn contains(&self, key: &DbKey) -> Result<bool, ChainStorageError>;
+    fn get_mmr_root(&self, tree: MmrTree) -> Result<HashOutput, ChainStorageError>;
+}
+
+// Private macro that pulls out all the boiler plate of extracting a DB query result from its variants
+macro_rules! fetch {
+    ($self:ident, $key_val:expr, $key_var:ident) => {{
+        let key = DbKey::$key_var($key_val);
+        match $self.db.get(&key) {
+            Ok(None) => Ok(None),
+            Ok(Some(DbValue::$key_var(k))) => Ok(Some(*k)),
+            Ok(Some(other)) => unexpected_result(key, other),
+            Err(e) => log_error(key, e),
+        }
+    }};
+}
+
+pub struct BlockChainDatabase<T>
+where T: BlockchainBackend
+{
+    metadata: Arc<RwLock<ChainMetadata>>,
+    db: Arc<T>,
+}
+
+impl<T> BlockChainDatabase<T>
```

Comment:
It might be an idea to call this BlockchainService or BlockchainManager, so that it's not confused with database backends

Reply:
Personally, I prefer the status quo. 
FWIW I had `BlockchainService`  earmarked for the futures-based service that delivers blockchain-related data.